xonsh/events.py

```diff
@@ -0,0 +1,169 @@
+        handler.__validator = None
+        self.add(handler)
+
+        def validator(vfunc):
+            """
+            Adds a validator function to a handler to limit when it is considered.
+            """
+            handler.__validator = vfunc
+        handler.validator = validator
+
+        return handler
+
+    def _filterhandlers(self, handlers, *pargs, **kwargs):
+        """
+        Helper method for implementing classes. Generates the handlers that pass validation.
+        """
+        for handler in handlers:
+            if handler.__validator is not None and not handler.__validator(*pargs, **kwargs):
+                continue
+            yield handler
+
+    @abc.abstractmethod
+    def fire(self, *pargs, **kwargs):
+        """
+        Fires an event, calling registered handlers with the given arguments.
+        """
+
+
+class Event(AbstractEvent):
+    """
+    A given event that handlers can register against.
+
+    Acts as a ``set`` for registered handlers.
+
+    Note that ordering is never guaranteed.
+    """
+    # Wish I could just pull from set...
+    def __init__(self):
+        self._handlers = set()
+
+    def __len__(self):
+        return len(self._handlers)
+
+    def __contains__(self, item):
+        return item in self._handlers
+
+    def __iter__(self):
+        yield from self._handlers
+
+    def add(self, item):
+        self._handlers.add(item)
+
+    def discard(self, item):
+        self._handlers.discard(item)
+
+    def fire(self, *pargs, **kwargs):
+        """
+        Fires each event, returning a non-unique iterable of the results.
+        """
+        vals = []
+        for handler in self._filterhandlers(self._handlers, *pargs, **kwargs):
+            try:
+                rv = handler(*pargs, **kwargs)
+            except Exception:
+                print("Exception raised in event handler; ignored.", file=sys.stderr)
+                traceback.print_exc()
+            else:
+                vals.append(rv)
+        return vals
+
+
+class LoadEvent(AbstractEvent):
+    """
+    A kind of event in which each handler is called exactly once.
+    """
+    def __init__(self):
+        self._fired = set()
+        self._unfired = set()
+        self._hasfired = False
+
+    def __len__(self):
+        return len(self._fired) + len(self._unfired)
+
+    def __contains__(self, item):
+        return item in self._fired or item in self._unfired
+
+    def __iter__(self):
+        yield from self._fired
+        yield from self._unfired
+
+    def add(self, item):
+        self._fired.add(item)
+
+    def discard(self, item):
+        self._fired.discard(item)
+        self._unfired.discard(item)
+
+    def fire(self, *pargs, **kwargs):
+        raise NotImplementedError("See #1550")
+
+
+class EventManager:
+    """
+    Container for all events in a system.
+
+    Meant to be a singleton, but doesn't enforce that itself.
```

Comment:
Having actual new instances helps ensure clean environments for tests.

It's the `Event` instances do all the real work anyway, so having multiple `EventManagers` just makes access weird. It shouldn't cause anything to actually break.